fdbrpc/AsyncFileNonDurable.actor.h

```diff
@@ -267,6 +267,10 @@ class AsyncFileNonDurable final : public IAsyncFile, public ReferenceCounted<Asy
 			Future<Void> deleteFuture = deleteFile(this);
 			if (!deleteFuture.isReady())
 				filesBeingDeleted[filename] = deleteFuture;
+		} else if (isSoleOwner()) {
+			// isSoleOwner is a bit confusing here. What we mean is that the openFiles map is the sole owner. If we
+			// remove the file from the map to make sure it gets closed.
+			g_simulator.getCurrentProcess()->machine->openFiles.erase(filename);
```

Comment:
I was told once upon a time (I think by @sfc-gh-satherton) that not closing Files was intentional, as it then simulation tests the situation where we write to a file, close it, later (re-)open it, and can read the written data, but that written data isn't durable on disk.

When I had looked at this before, I sketched out something in #2766 that would close file descriptors but not Files. I don't have the memory of the code anymore though to actually remember if this code actually does reproduce/test the situation I described above though.

Reply:
Thank you, I was having trouble remembering this but yes this was my understanding of why files are kept open after close.

In the real world, I think this read-your-own-non-durable-writes-across-reopen behavior can happen with both direct and cached IO.  With a file descriptor you can write, close, open, read, and see data that is not yet fsync'd due to the OS block cache or the disk's memory cache.  With direct IO, I think the latter still applies.

I think the reason to simulate the behavior is to catch durability errors like this:

1.  Write to file A
2. Close file A
3. Reopen file A to recover some state
4. Read file A, assuming all data read is durable
5. Write file B with data that references data in file A
6. Sync file B

and now upon recovery this two-file structure is broken.

I think there is still a fix possible here.  When a file is closed, if it has no unsync'd writes - and I think AsyncFileNonDurable might know this already - then it can be closed and removed from the open file map.

Reply:
can you please point me to the code that injects these kind of failures? Currently files are not closed (except for disk queue files after they got renamed -- which means the main point where we want to test something like the above this doesn't change behavior), so state is going to be available after step 2 -- even if the state has not yet been written to disk.

Reply:
Steve, Evan, and I looked closely into this. It doesn't seem that this code change will change current behavior except for the case where someone closes a file before writes are durable and then immediately opens it. So I pushed a change that will make sure the behavior doesn't change.